core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -718,21 +720,32 @@ protected Integer getLivenessCheckPort() {
 
     private void applyConfiguration(CreateContainerCmd createCommand) {
         HostConfig hostConfig = buildHostConfig();
-        createCommand.withHostConfig(hostConfig);
 
-        // Set up exposed ports (where there are no host port bindings defined)
-        ExposedPort[] portArray = exposedPorts.stream()
-                .map(ExposedPort::new)
-                .toArray(ExposedPort[]::new);
+        // PortBindings must contain:
+        //  * all exposed ports with a randomized host port (equivalent to -p CONTAINER_PORT)
+        //  * all exposed ports with a fixed host port (equivalent to -p HOST_PORT:CONTAINER_PORT)
+        Map<ExposedPort, PortBinding> allPortBindings = new HashMap<>();
+        // First collect all the randomized host ports from our 'exposedPorts' field
+        for (final Integer tcpPort : exposedPorts) {
+            ExposedPort exposedPort = new ExposedPort(tcpPort);
+            allPortBindings.put(exposedPort, new PortBinding(Ports.Binding.empty(), exposedPort));
+        }
+        // Next collect all the fixed host ports from our 'portBindings' field, overwriting any randomized ports so that
+        // we don't create two bindings for the same container port.
+        for (final String portBinding : portBindings) {
+            PortBinding parsedBinding = PortBinding.parse(portBinding);
+            allPortBindings.put(parsedBinding.getExposedPort(), parsedBinding);
+        }
 
-        createCommand.withExposedPorts(portArray);
+        List<PortBinding> finalBindings = new ArrayList<>(allPortBindings.values());
+        hostConfig.withPortBindings(finalBindings);
 
-        // Set up exposed ports that need host port bindings
-        PortBinding[] portBindingArray = portBindings.stream()
-                .map(PortBinding::parse)
-                .toArray(PortBinding[]::new);
+        // Next, ExposedPorts must be set up to publish all of the above ports, both randomized and fixed.
+        // Collect all of the exposed ports for publication
+        final List<ExposedPort> finalExposedPorts = new ArrayList<>(allPortBindings.keySet());
+        createCommand.withExposedPorts(finalExposedPorts);
```

Comment:
Ditto: https://github.com/testcontainers/testcontainers-java/pull/4122/files#r638623199

```suggestion
        createCommand.withExposedPorts(new ArrayList<>(allPortBindings.keySet()));
```